SeamCarver.js

```diff
@@ -45,7 +77,143 @@ class SeamCarver {
+            (xpost[RED] - xant[RED])*(xpost[RED] - xant[RED]) +
+            (xpost[GREEN] - xant[GREEN])*(xpost[GREEN] - xant[GREEN]) +
+            (xpost[BLUE] - xant[BLUE])*(xpost[BLUE] - xant[BLUE]) +
+            (ypost[RED] - yant[RED])*(ypost[RED] - yant[RED]) +
+            (ypost[GREEN] - yant[GREEN])*(ypost[GREEN] - yant[GREEN]) +
+            (ypost[BLUE] - yant[BLUE])*(ypost[BLUE] - yant[BLUE])
+        );
+    }
+
+    /**
+     * Calculate energy_matrix information for pixel x,y.
+     * Assumes x and y in range.
+     */
+    recalculate(x, y) {
+        var energy_cell = {};
+
+        energy_cell.energy = this.energy(x, y);
+        energy_cell.vminsum = Number.POSITIVE_INFINITY;
 
+        // last row
+        if (y >= this.height-1) {
+            energy_cell.vminsum = energy_cell.energy;
+            energy_cell.minx = 0;
+        } else {
+            var cursum = 0;
+            var curminx = 0;
+
+            // below left
+            if (x - 1 >= 0) {
+                energy_cell.vminsum = this.energy_matrix[x - 1][y + 1].vminsum + energy_cell.energy;
+                energy_cell.minx = x - 1;
+            }
+
+            // below
+            cursum = this.energy_matrix[x][y + 1].vminsum + energy_cell.energy;
+            if (cursum < energy_cell.vminsum) {
+                energy_cell.vminsum = cursum;
+                energy_cell.minx = x;
+            }
+
+            // below right
+            if (x + 1 < this.width) {
+                cursum = this.energy_matrix[x + 1][y + 1].vminsum + energy_cell.energy;
+                if (cursum < energy_cell.vminsum) {
+                    energy_cell.vminsum = cursum;
+                    energy_cell.minx = x + 1;
+                }
+            }
+        }
+
+        return energy_cell;
+    }
+
+    /**
+     * Iterate from bottom to top. For each pixel calculate:
+     *     * The energy for the pixel.
+     *     * From the three pixels below the current pixel, calculate the
+     *       `min_x` pixel. The `min_x` pixel is the pixel with the smallest
+     *       cumulative energy (defined below).
+     *     * Set the cumulative energy for this pixel as the energy of this
+     *       pixel plus the cumulative energy of th `min_x` pixel.
+     *
+     * The cumulative energy of the pixels in the bottom row is simply its own
+     * energy.
+     *
+     */
+    createEnergyMatrix() {
+        // This has to be reverse order (bottom to top)
+        for (var y = this.height; y >= 0; y--) {
+            // This can be in any order ...
+            for (var x = 0; x < this.width; x++) {
+                this.energy_matrix[x][y] = this.recalculate(x,y);
+            }
+        }
+    }
+
+    /**
+     * Backtrack from smallest on first row to choosing always smallest child.
+     *
+     */
+    findVerticalSeam() {
+        var vseam = [];
+
+        var xminsum = 0;
+        var vminsum = Number.POSITIVE_INFINITY;
+
+        // Find smallest sum on first row
```

Comment:
We could keep track of the min pixel on the first row in the `createEnergyMatrix` method